src/realm/object-store/c_api/conversion.hpp

```diff
@@ -25,6 +25,10 @@ static inline realm_string_t to_capi(StringData data)
     return realm_string_t{data.data(), data.size()};
 }
 
+// Because this is often used as `return to_capi(...);` it is dangerous to pass a temporary string here. If you really
+// need to and know it is correct (eg passing to a C callback), you can explicitly create the StringData wrapper.
+realm_string_t to_capi(const std::string&& str) = delete; // temporary std::string would dangle.
```

Comment:
I confirmed that this would catch my `return to_capi(cond ? constStringRef : "")` bug where that expression produced a temporary `std::string`. I usually don't love banning rvalue overloads because that prevents safe usage as much as it unsafe, but because A) `to_capi()` is often used in return position and B) it is easy to work around where it is known to be safe, it seems worth it here. If you disagree, I can revert the commit to add this overload.

The line comment *is* redundant with the comment above, however I like having it because it will show up in  compiler error messages, while the above comment won't.